skimage/segmentation/_join.py

```diff
@@ -116,26 +117,172 @@ def relabel_sequential(label_field, offset=1):
+    input_vals : array of int, shape (N,)
+        The values to map from.
+    output_vals : array, shape (N,)
+        The values to map to.
+    out: array, same shape as `input_arr`
+        The output array. Will be created if not provided. It should
+        have the same dtype as `output_vals`.
+
+    Returns
+    -------
+    out : array, same shape as `input_arr`
+        The array of mapped values.
+    """
+
+    # We ravel the input array for simplicity of iteration in Cython:
+    orig_shape = input_arr.shape
+    # NumPy docs for `np.ravel()` says:
+    # "When a view is desired in as many cases as possible, 
+    # arr.reshape(-1) may be preferable."
+    input_arr = input_arr.reshape(-1)
+    if out is None:
+        out = np.empty_like(input_arr, dtype=output_vals.dtype)
+    elif not (out.flags['FORC'] or out.ndim == 1):
+        raise ValueError(
+            'If out array is provided, it should be either contiguous '
+            'or 1-dimensional.'
+        )
+
+    _map_array(input_arr, out, input_vals, output_vals)
+    return out.reshape(orig_shape)
+
+
+class ArrayMap:
+    """Class designed to mimic mapping by NumPy array indexing.
+
+    This class is designed to replicate the use of NumPy arrays for mapping
+    values with indexing:
+
+    >>> values = np.array([0.25, 0.5, 1.0])
+    >>> indices = np.array([[0, 0, 1], [2, 2, 1]])
+    >>> values[indices]
+    array([[0.25, 0.25, 0.5 ],
+           [1.  , 1.  , 0.5 ]])
+
+    The issue with this indexing is that you need a very large ``values``
+    array if the values in the ``indices`` array are large.
+
+    >>> values = np.array([0.25, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1.0])
+    >>> indices = np.array([[0, 0, 10], [0, 10, 10]])
+    >>> values[indices]
+    array([[0.25, 0.25, 1.  ],
+           [0.25, 1.  , 1.  ]])
+
+    Using this class, the approach is similar, but there is no need to
+    create a large values array:
+
+    >>> in_indices = np.array([0, 10])
+    >>> out_values = np.array([0.25, 1.0])
+    >>> values = ArrayMap(in_indices, out_values)
+    >>> indices = np.array([[0, 0, 10], [0, 10, 10]])
+    >>> values[indices]
+    array([[0.25, 0.25, 1.  ],
+           [0.25, 1.  , 1.  ]])
+
+    Parameters
+    ----------
+    in_values : array of int, shape (N,)
+        The source values from which to map.
+    out_values : array, shape (N,)
+        The destination values from which to map.
+    """
+    def __init__(self, in_values, out_values):
+        self.in_values = in_values
+        self.out_values = out_values
+        self._max_lines = 4
+
+    def __array__(self, dtype=None):
+        """Return an array that behaves like the arraymap when indexed.
+        
+        This array can be very large: it is the size of the largest value
+        in the ``in_vals`` array, plus one.
+        """
+        if dtype is None:
+            dtype = self.out_values.dtype
+        output = np.zeros(np.max(self.in_values) + 1, dtype=dtype)
```

Comment:
This array may be huge if `np.max(self.in_values)` is large....

Reply:
yes, but that is what this PR is fixing... We are only providing `__array__` as a compatibility shim for existing users of array indexing. ie this PR strictly *decreases* the surface area of this problem.

Reply:
OK, I see...

Can't we consider using a `return_sparse_maps` (or equivalent) argument:
- if `False`, the function returns the dense `fw_map` and `inv_map` as actually,
- if `True`, the sparse version of `fw_map` and `inv_map` in any of the `scipy.sparse` available sparse matrix formats.

and in the documentation we show how remapping can be made with both options:

```python
>>> from scipy import sparse as Sp
>>> from skimage.segmentation import relabel_sequential
>>> # Dense version
>>> label_field = np.array([1, 1, 5, 5, 8, 99, 42])
>>> relab, fw, inv = relabel_sequential(label_field)
>>> (fw[label_field] == relab).all()
True
>>> (inv[relab] == label_field).all()
True
>>> # Sparse version
>>> sp_fw = Sp.lil_matrix(fw)
>>> sp_inv = Sp.lil_matrix(inv)
>>> (sp_fw[0, label_field] == relab).all()
True
>>> (sp_inv[0, relab] == label_field).all()
True
```

Reply:
I don't think that is necessary. The `__array__` method of  `ArrayMap` is really only needed for backwards compatibiliy. If the memory wasn't a problem in theses historic cases it should not be now. In the future, the forward and backward mappings should simply be invoked as if it were fancy indexing (the `__getitem__` method takes care of that and if memory-efficient) or like a function call (`__call__`).

BTW, I had looked whether the original approach of using an array as LUT could be retained with `scipy.sparse` arrays before setting out on using Cython. I hadn't tried it but IIRC I read in the documentation that the sparse arrays don't support fancy indexing (or need to be converted to numpy arrays first).

Reply:
@rfezzani I still prefer the ArrayMap solution, for the following reasons:

(1) the sparse relabeling is actually much slower (>3x) even than the Cython code, which is already a bit slower than the NumPy code for moderately-sized images.
(2) we are trying to move away from `return_*` arguments
(3) we will need to deprecate the current default in favour of a new default, which is a pain
(4) the code is clever (and I do like fancy uses of csr matrices, I do it all the time!), but it is less readable/understandable than the Cython code, which is a straightforward loop
(5) @VolkerH is right, the fancy indexing with the sparse arrays only works for 1D arrays, so relabeling with sparse needs to go through a ravel and a reshape.

The dtype of the matrices also needs to be tweaked, too.

Reply:
No problem, but:

(1) I will have to do some more experiments to be convinced =p,
(2) I agree, the `return_*` arguments are not a good choice, but a better wording is may be possible,
(3) No deprecation is needed: if you see my [comment](https://github.com/scikit-image/scikit-image/pull/4612#discussion_r417581951), the actual behavior can be kept as default and the sparse arrays advised only for advanced use, when memory can be an issue
(4) this is subjective :)
(5) that's a major drawback, but if we consider that sparse matrices are only for advanced use, just few lines of documentation can suffices =p

The dtype is not much problematic with matrices then with arrays like it is the case in the actual implementation.

BTW, what prevents `relabel_sequence` from processing large values is the huge memory use that it implies, so if the `__array__` method is required and called in the `ArrayMap` approach, I wonder if the issue is addressed.

Reply:
As you mentioned, this is partly subjective, but just point 3 is a deal-breaker for me:

3. If I write a program the uses `relabel_sequential` I don't necessarily know what sort of maximum labels someone might provide as input data. Therefore I would have to add code that finds the maximum label for each provided input image and then either set an arbitrary threshold or check how much memory is available to decide whether to call it with the option `return_sparse` or not. 

In my opinion, it should simply work independently of the maximum label value without having to provide such an option.